components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt

```diff
@@ -0,0 +1,413 @@
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.ContentPermission.VALUE_ALLOW
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.ContentPermission.VALUE_DENY
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.ContentPermission.VALUE_PROMPT
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_AUTOPLAY_AUDIBLE
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_AUTOPLAY_INAUDIBLE
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_DESKTOP_NOTIFICATION
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_GEOLOCATION
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_MEDIA_KEY_SYSTEM_ACCESS
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_PERSISTENT_STORAGE
+import org.mozilla.geckoview.GeckoSession.PermissionDelegate.PERMISSION_TRACKING
+import org.mozilla.geckoview.StorageController
+import org.mozilla.geckoview.StorageController.ClearFlags
+
+/**
+ * A storage to save [SitePermissions] using GeckoView APIs.
+ */
+@Suppress("LargeClass")
+class GeckoSitePermissionsStorage(
+    runtime: GeckoRuntime,
+    private val onDiskStorage: SitePermissionsStorage
+) : SitePermissionsStorage {
+
+    private val geckoStorage: StorageController = runtime.storageController
+    private val mainScope = CoroutineScope(Dispatchers.Main)
+
+    /*
+    * At the moment gecko view doesn't handle temporary permission,
+    * we have to store them in memory, clear them manually,
+    * until we have an API for it see:
+    * https://bugzilla.mozilla.org/show_bug.cgi?id=1710447
+    * */
+    @VisibleForTesting
+    internal val geckoTemporaryPermissions = mutableListOf<ContentPermission>()
+
+    override suspend fun save(sitePermissions: SitePermissions, request: PermissionRequest?) {
+        val geckoSavedPermissions = updateGeckoPermissionIfNeeded(sitePermissions, request)
+        onDiskStorage.save(geckoSavedPermissions, request)
+    }
+
+    override fun saveTemporary(request: PermissionRequest?) {
+        if (request is GeckoPermissionRequest.Content) {
+            geckoTemporaryPermissions.add(request.geckoPermission)
+        }
+    }
+
+    override fun clearTemporaryPermissions() {
+        geckoTemporaryPermissions.forEach {
+            geckoStorage.setPermission(it, VALUE_PROMPT)
+        }
+        geckoTemporaryPermissions.clear()
+    }
+
+    override suspend fun update(sitePermissions: SitePermissions) {
+        val updatedPermission = updateGeckoPermissionIfNeeded(sitePermissions)
+        onDiskStorage.update(updatedPermission)
+    }
+
+    override suspend fun findSitePermissionsBy(origin: String, includeTemporary: Boolean): SitePermissions? {
+        /**
+         * GeckoView ony persists [GeckoPermissionRequest.Content] other permissions like
+         * [GeckoPermissionRequest.Media], we have to store them ourselves.
```

Comment:
Ah, this may be the answer to my other question. There are still some left we need to save ourselves..?

Reply:
Is there anything planned for the future?

Reply:
As far as I know far GV doesn't plan store media permissions in the near feature :(

Reply:
Bummer! Maybe we can file it and see if they would be open to it? :)